### DISCUSSION TITLE
ctsem for thesis in psychology, stanct approach

### DISCUSSION THREAD

**Discussion by ViktoriaJustus** [cdriveraus/ctsem#14] [Q&A]
Hello Charles,

thanks for the response in the other forum. 

So shortly to the background again:
I am writing my masters thesis in organizational psychology and I collected some data by means of the diary method (15 days with 2 entries per day, 73 individuals, about different 10 variables).

I had a look at the stanct model specification and tried to use it for my data instead of the omx approach. I took the most simple one for the first runs as this aproach is new to me. So this is the code for my 1. Hypothesis which assumes that Feedback (DF) triggers Reflection (AR), and Reflection also influences Feedback over time.

#load packages
library("ctsem")
library("Rcpp")
library("OpenMx")
library("ctsemOMX")
library("MASS")

#load data
long_alle_Daten_H1 <- read.csv2("...long_alle_Daten_H1.csv", TRUE)

#have a look at data
head(long_alle_Daten_H1)
long_alle_Daten_H1

#generate a ctStanModel relying heavily on defaults
sm<-ctModel(type='stanct',id='id',time = 'time',
  manifestNames = c('DF','AR'),
  MANIFESTVAR=0, #disables measurement error for this example
  LAMBDA=matrix(c(1, 0, 0, 1), nrow=2, ncol=2))

#Fit using maximum likelihood:
sf=ctStanFit(long_alle_Daten_H1, ctstanmodel = sm,cores=6)

#use summary functions:
s=summary(sf, timeinterval = 1)
print(s)

So the output looks quite different to the omx output which is a bit confusing. I have the following issues concerning that: 
1) I cannot identify any beta coefficients which show the relationship of the two variables. 
2) There are z-Values, but I cannot find p-values to report. 
3) Is there a possibility to receive model fit statistics (CFI, Chi squared, RMSEA, SRMR)?

Moreover, I have two more general questions: 
1) I have lots of missing values in the variables and I have read that is no big deal with this approach. However, would it be better to impute values for the missing data?
2) I have a moderation hypothesis, so I wonder if it is possible to code a moderation with ctsem. 

Thanks in advance!
I appreciate your help very much :)

**Comment by cdriveraus** [comment]
1. The two processes are related in a variety of different ways under the default spec. They will have correlated manifest means (individual specific intercepts / traits), which will also be correlated with the initial values of the latent processes (t0means).  Unpredictable fluctuations in the latent processes will have a covariance as shown by the DIFFUSIONcov matrix, and determined by the off diagonals of the DIFFUSION matrix. Finally, the latent processes influence / predict each other according to the off diagonals of the DRIFT matrix. Precisely which you're interested in depends on the research question/s and how much you're willing to believe in the model. I'd suggest viewing the model spec in matrix form using `ctModelLatex(mymodel)`

2. p values from model comparisons can be obtained with the ctChisqTest() function, pass it two fit objects. For the parameters themselves, hopefully the confidence intervals are enough, or you'll need to find a way to compute what you want from the samples found in the $stanfit$transformedpars sub-object, possibly using the ctAddSamples() function to obtain more samples for precision. 

3. I'm not a big believer in such stats and the appropriate basis of comparison in many continuous time / nonlinear cases is not very clear, but if you fit one or more models using the ctFitMultiModel function, it will also output a range of summary / model fit checking type plots and provide some comparison stats of the models. For some models the summary output can be slow. You can also use the not yet exported function ctSaturatedFit() to obtain the likelihood of a saturated covariance matrix for a semi arbitrary structuring of the data used in the model, like so :

```
myfit <- ctsem:::ctSaturatedFit(myfit,oos=FALSE)
print(myfit$stanfit$saturated$ll)
stats::pchisq(q =  2*diff(c(myfit$stanfit$optimfit$value, myfit$stanfit$saturated$ll)),
  df = diff(c(length(myfit$stanfit$rawest),myfit$stanfit$saturated$npars)),lower.tail = FALSE)
```

1. Missing values on the manifest variables are better off not imputed unless you have a particularly well developed imputation model. Missings on covariates are imputed by default and a warning shown -- you can definitely do a better job here. Missing values on the time variable are also not possible, so need to be provided in some way (or use the discrete time models). 

2. One process moderates a relationship between two others? Or the relationship is moderated by a covariate (time independent predictor)? Both are possible, the latter is much simpler. For the latter, any covariates will by default influence the relationship, and you may simply find there is too much freedom / too many parameters estimated so some effects need to be set to zero (see the ctDocs() documentation). For the case that it is a latent process doing the moderating, you are in nonlinear territory and should take a look at that section of the docs, one of the off diagonals in the DRIFT matrix will need to look something like: 'relationbase + moderationeffect * moderatingprocess' and you'll need `PARS=c('relationbase','moderationeffect')` in your model specification.

**Reply by cdriveraus** [reply]
I forgot: assuming zero measurement error is a pretty bad idea in most circumstances. will bias coefficients / generate spurious looking relations and limit the models ability to account for the most important lag time.

**Comment by ViktoriaJustus** [comment]
Hello Charles, 

thanks for your support last time :) 

Is there a possibility to figure out the maximum beta coefficient the the time lag for that? Attached is an example plot of the standardised cross regression. For example here I would like to know the highest point of the grey line. 

![Standardisierte Crossregression - Reflexion (Rges)  Feedback (Fges)](https://user-images.githubusercontent.com/84229077/119240482-1213d700-bb50-11eb-9f1e-d6dc3e22907a.png)

Thanks in advance!

**Reply by cdriveraus** [reply]
just use the ctStanDiscretePars function to compute the sequence for small time intervals around where you see the maximum -- you can specify the time sequence as an argument and if you leave plot=FALSE it will output the results. There might be an analytic solution but I don't know it :)
